Source/Charts/Data/Implementations/Standard/ChartDataSet.swift

```diff
@@ -260,98 +213,57 @@ open class ChartDataSet: ChartBaseDataSet
         closestToY yValue: Double,
         rounding: ChartDataSetRounding) -> Int
     {
-        var low = startIndex
-        var high = endIndex - 1
-        var closest = high
-        
-        while low < high
-        {
-            let m = (low + high) / 2
-            
-            let d1 = self[m].x - xValue
-            let d2 = self[m + 1].x - xValue
-            let ad1 = abs(d1), ad2 = abs(d2)
-            
-            if ad2 < ad1
-            {
-                // [m + 1] is closer to xValue
-                // Search in an higher place
-                low = m + 1
-            }
-            else if ad1 < ad2
+        var closest = partitioningIndex { $0.x >= xValue }
+        guard closest < endIndex else { return closest }
+
+        let closestXValue = self[closest].x
+
+        switch rounding {
+        case .up:
+            // If rounding up, and found x-value is lower than specified x, and we can go upper...
+            if closestXValue < xValue && closest < index(before: endIndex)
             {
-                // [m] is closer to xValue
-                // Search in a lower place
-                high = m
+                formIndex(after: &closest)
             }
-            else
+
+        case .down:
+            // If rounding down, and found x-value is upper than specified x, and we can go lower...
+            if closestXValue > xValue && closest > startIndex
             {
-                // We have multiple sequential x-value with same distance
-                
-                if d1 >= 0.0
-                {
-                    // Search in a lower place
-                    high = m
-                }
-                else if d1 < 0.0
-                {
-                    // Search in an higher place
-                    low = m + 1
-                }
+                formIndex(before: &closest)
             }
-            
-            closest = high
+
+        case .closest:
+            break
         }
-        
-        if closest != -1
```

Comment:
The only case where `closest == -1` in the old implementation is if the dataset is empty (`low - high == -1`). This should probably be checked immediately, rather than later on.

Returning `-1` is very non-swift, which is why I've elected to drop that behaviour from everywhere I've found it in 4.0. 

`partitioningIndex` returns `endIndex` if no index is found in `startIndex..<endIndex`